File: smalivm/src/main/java/org/cf/smalivm/opcode/OpCreator.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntObjectMap;

import org.cf.smalivm.VirtualMachine;
import org.jf.dexlib2.Opcode;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.Instruction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class OpCreator {

    @SuppressWarnings("unused")
    private static final Logger log = LoggerFactory.getLogger(OpCreator.class.getSimpleName());
    private final VirtualMachine vm;
    private final TIntObjectMap<MethodLocation> addressToLocation;

    public OpCreator(VirtualMachine vm, TIntObjectMap<MethodLocation> addressToLocation) {
        this.vm = vm;
        this.addressToLocation = addressToLocation;
    }

    public Op create(MethodLocation location) {
        Instruction instruction = location.getInstruction();
        Opcode opcode = instruction.getOpcode();
        OpFactory opFactory = getOpFactory(opcode);

        return opFactory.create(location, addressToLocation, vm);
    }

    @SuppressWarnings("incomplete-switch")
    private static OpFactory getOpFactory(Opcode op) {
        OpFactory opFactory = null;
        switch (op) {
            case ADD_DOUBLE:
            case ADD_DOUBLE_2ADDR:
            case ADD_FLOAT:
            case ADD_FLOAT_2ADDR:
            case ADD_INT:
            case ADD_INT_2ADDR:
            case ADD_INT_LIT16:
            case ADD_INT_LIT8:
            case ADD_LONG:
            case ADD_LONG_2ADDR:
            case AND_INT:
            case AND_INT_2ADDR:
            case AND_INT_LIT16:
            case AND_INT_LIT8:
            case AND_LONG:
            case AND_LONG_2ADDR:
            case DIV_DOUBLE:
            case DIV_DOUBLE_2ADDR:
            case DIV_FLOAT:
            case DIV_FLOAT_2ADDR:
            case DIV_INT:
            case DIV_INT_2ADDR:
            case DIV_INT_LIT16:
            case DIV_INT_LIT8:
            case DIV_LONG:
            case DIV_LONG_2ADDR:
            case MUL_DOUBLE:
            case MUL_DOUBLE_2ADDR:
            case MUL_FLOAT:
            case MUL_FLOAT_2ADDR:
            case MUL_INT:
            case MUL_INT_2ADDR:
            case MUL_INT_LIT16:
            case MUL_INT_LIT8:
            case MUL_LONG:
            case MUL_LONG_2ADDR:
            case OR_INT:
            case OR_INT_2ADDR:
            case OR_INT_LIT16:
            case OR_INT_LIT8:
            case OR_LONG:
            case OR_LONG_2ADDR:
            case REM_DOUBLE:
            case REM_DOUBLE_2ADDR:
            case REM_FLOAT:
            case REM_FLOAT_2ADDR:
            case REM_INT:
            case REM_INT_2ADDR:
            case REM_INT_LIT16:
            case REM_INT_LIT8:
            case REM_LONG:
            case REM_LONG_2ADDR:
            case RSUB_INT:
            case RSUB_INT_LIT8:
            case SHL_INT:
            case SHL_INT_2ADDR:
            case SHL_INT_LIT8:
            case SHL_LONG:
            case SHL_LONG_2ADDR:
            case SHR_INT:
            case SHR_INT_2ADDR:
            case SHR_INT_LIT8:
            case SHR_LONG:
            case SHR_LONG_2ADDR:
            case SUB_DOUBLE:
            case SUB_DOUBLE_2ADDR:
            case SUB_FLOAT:
            case SUB_FLOAT_2ADDR:
            case SUB_INT:
            case SUB_INT_2ADDR:
            case SUB_LONG:
            case SUB_LONG_2ADDR:
            case USHR_INT:
            case USHR_INT_2ADDR:
            case USHR_INT_LIT8:
            case USHR_LONG:
            case USHR_LONG_2ADDR:
            case XOR_INT:
            case XOR_INT_2ADDR:
            case XOR_INT_LIT16:
            case XOR_INT_LIT8:
            case XOR_LONG:
            case XOR_LONG_2ADDR:
                opFactory = new BinaryMathOpFactory();
                break;

            case AGET:
            case AGET_BOOLEAN:
            case AGET_BYTE:
            case AGET_CHAR:
            case AGET_OBJECT:
            case AGET_SHORT:
            case AGET_WIDE:
                opFactory = new AGetOpFactory();
                break;

            case APUT:
            case APUT_BOOLEAN:
            case APUT_BYTE:
            case APUT_CHAR:
            case APUT_OBJECT:
            case APUT_SHORT:
            case APUT_WIDE:
                opFactory = new APutOpFactory();
                break;

            case ARRAY_LENGTH:
                opFactory = new ArrayLengthOpFactory();
                break;

            case ARRAY_PAYLOAD:
                opFactory = new FillArrayDataPayloadOpFactory();
                break;

            case CHECK_CAST:
                opFactory = new CheckCastOpFactory();
                break;

            case CMPG_DOUBLE:
            case CMPG_FLOAT:
            case CMPL_DOUBLE:
            case CMPL_FLOAT:
            case CMP_LONG:
                opFactory = new CmpOpFactory();
                break;

            case CONST:
            case CONST_16:
            case CONST_4:
            case CONST_CLASS:
            case CONST_HIGH16:
            case CONST_STRING:
            case CONST_STRING_JUMBO:
            case CONST_WIDE:
            case CONST_WIDE_16:
            case CONST_WIDE_32:
            case CONST_WIDE_HIGH16:
                opFactory = new ConstOpFactory();
                break;

            case DOUBLE_TO_FLOAT:
            case DOUBLE_TO_INT:
            case DOUBLE_TO_LONG:
            case FLOAT_TO_DOUBLE:
            case FLOAT_TO_INT:
            case FLOAT_TO_LONG:
            case INT_TO_BYTE:
            case INT_TO_CHAR:
            case INT_TO_DOUBLE:
            case INT_TO_FLOAT:
            case INT_TO_LONG:
            case INT_TO_SHORT:
            case LONG_TO_DOUBLE:
            case LONG_TO_FLOAT:
            case LONG_TO_INT:
            case NEG_DOUBLE:
            case NEG_FLOAT:
            case NEG_INT:
            case NEG_LONG:
            case NOT_INT:
            case NOT_LONG:
                opFactory = new UnaryMathOpFactory();
                break;

            case FILLED_NEW_ARRAY:
            case FILLED_NEW_ARRAY_RANGE:
                opFactory = new FilledNewArrayOpFactory();
                break;

            case FILL_ARRAY_DATA:
                opFactory = new FillArrayDataOpFactory();
                break;

            case GOTO:
            case GOTO_16:
            case GOTO_32:
                opFactory = new GotoOpFactory();
                break;

            case IF_EQ:
            case IF_GE:
            case IF_GT:
            case IF_LE:
            case IF_LT:
            case IF_NE:
            case IF_EQZ:
            case IF_GEZ:
            case IF_GTZ:
            case IF_LEZ:
            case IF_LTZ:
            case IF_NEZ:
                opFactory = new IfOpFactory();
                break;

            case IGET:
            case IGET_BOOLEAN:
            case IGET_BYTE:
            case IGET_CHAR:
            case IGET_OBJECT:
            case IGET_SHORT:
            case IGET_WIDE:
                opFactory = new IGetOpFactory();
                break;

            case IPUT:
            case IPUT_BOOLEAN:
            case IPUT_BYTE:
            case IPUT_CHAR:
            case IPUT_OBJECT:
            case IPUT_SHORT:
            case IPUT_WIDE:
                opFactory = new IPutOpFactory();
                break;

            case INSTANCE_OF:
                opFactory = new InstanceOfOpFactory();
                break;

            case INVOKE_DIRECT:
            case INVOKE_INTERFACE:
            case INVOKE_STATIC:
            case INVOKE_SUPER:
            case INVOKE_VIRTUAL:
            case INVOKE_DIRECT_RANGE:
            case INVOKE_INTERFACE_RANGE:
            case INVOKE_STATIC_RANGE:
            case INVOKE_SUPER_RANGE:
            case INVOKE_VIRTUAL_RANGE:
                opFactory = new InvokeOpFactory();
                break;

            case MONITOR_ENTER:
                opFactory = new MonitorEnterOpFactory();
                break;
            case MONITOR_EXIT:
                opFactory = new MonitorExitOpFactory();
                break;

            case MOVE:
            case MOVE_16:
            case MOVE_FROM16:
            case MOVE_OBJECT:
            case MOVE_OBJECT_16:
            case MOVE_OBJECT_FROM16:
            case MOVE_WIDE:
            case MOVE_WIDE_16:
            case MOVE_WIDE_FROM16:
            case MOVE_EXCEPTION:
            case MOVE_RESULT:
            case MOVE_RESULT_OBJECT:
            case MOVE_RESULT_WIDE:
                opFactory = new MoveOpFactory();
                break;

            case NEW_ARRAY:
                opFactory = new NewArrayOpFactory();
                break;

            case NEW_INSTANCE:
                opFactory = new NewInstanceOpFactory();
                break;

            case NOP:
                opFactory = new NopOpFactory();
                break;

            case PACKED_SWITCH:
            case SPARSE_SWITCH:
                opFactory = new SwitchOpFactory();
                break;

            case PACKED_SWITCH_PAYLOAD:
            case SPARSE_SWITCH_PAYLOAD:
                opFactory = new SwitchPayloadOpFactory();
                break;

            case RETURN:
            case RETURN_WIDE:
            case RETURN_OBJECT:
                opFactory = new ReturnOpFactory();
                break;

            case RETURN_VOID:
                opFactory = new ReturnVoidOpFactory();
                break;

            case SGET:
            case SGET_BOOLEAN:
            case SGET_BYTE:
            case SGET_CHAR:
            case SGET_OBJECT:
            case SGET_SHORT:
            case SGET_WIDE:
                opFactory = new SGetOpFactory();
                break;

            case SPUT:
            case SPUT_BOOLEAN:
            case SPUT_BYTE:
            case SPUT_CHAR:
            case SPUT_OBJECT:
            case SPUT_SHORT:
            case SPUT_WIDE:
                opFactory = new SPutOpFactory();
                break;

            case THROW:
                opFactory = new ThrowOpFactory();
                break;
        }

        if (opFactory == null) {
            throw new RuntimeException("Unimplemented opcode: " + op);
        }

        return opFactory;
    }
}

File: smalivm/src/main/java/org/cf/smalivm/opcode/OpCreator.kt
package org.cf.smalivm.opcode

import ExceptionFactory
import org.cf.smalivm.type.ClassManager
import org.jf.dexlib2.Opcode
import org.jf.dexlib2.builder.MethodLocation
import org.slf4j.LoggerFactory

class OpCreator(
    private val addressToLocation: Map<Int, MethodLocation>,
    private val classManager: ClassManager,
    private val exceptionFactory: ExceptionFactory
) {
    fun create(location: MethodLocation): Op {
        val instruction = location.instruction!!
        val opcode = instruction.opcode
        val opFactory = getOpFactory(opcode)
        return opFactory.build(location, addressToLocation, classManager, exceptionFactory,,)
    }

    companion object {
        private val log = LoggerFactory.getLogger(OpCreator::class.java.simpleName)

        private fun getOpFactory(op: Opcode): OpFactory {
            return when (op) {
                Opcode.ADD_DOUBLE, Opcode.ADD_DOUBLE_2ADDR, Opcode.ADD_FLOAT, Opcode.ADD_FLOAT_2ADDR, Opcode.ADD_INT, Opcode.ADD_INT_2ADDR, Opcode.ADD_INT_LIT16, Opcode.ADD_INT_LIT8, Opcode.ADD_LONG, Opcode.ADD_LONG_2ADDR, Opcode.AND_INT, Opcode.AND_INT_2ADDR, Opcode.AND_INT_LIT16, Opcode.AND_INT_LIT8, Opcode.AND_LONG, Opcode.AND_LONG_2ADDR, Opcode.DIV_DOUBLE, Opcode.DIV_DOUBLE_2ADDR, Opcode.DIV_FLOAT, Opcode.DIV_FLOAT_2ADDR, Opcode.DIV_INT, Opcode.DIV_INT_2ADDR, Opcode.DIV_INT_LIT16, Opcode.DIV_INT_LIT8, Opcode.DIV_LONG, Opcode.DIV_LONG_2ADDR, Opcode.MUL_DOUBLE, Opcode.MUL_DOUBLE_2ADDR, Opcode.MUL_FLOAT, Opcode.MUL_FLOAT_2ADDR, Opcode.MUL_INT, Opcode.MUL_INT_2ADDR, Opcode.MUL_INT_LIT16, Opcode.MUL_INT_LIT8, Opcode.MUL_LONG, Opcode.MUL_LONG_2ADDR, Opcode.OR_INT, Opcode.OR_INT_2ADDR, Opcode.OR_INT_LIT16, Opcode.OR_INT_LIT8, Opcode.OR_LONG, Opcode.OR_LONG_2ADDR, Opcode.REM_DOUBLE, Opcode.REM_DOUBLE_2ADDR, Opcode.REM_FLOAT, Opcode.REM_FLOAT_2ADDR, Opcode.REM_INT, Opcode.REM_INT_2ADDR, Opcode.REM_INT_LIT16, Opcode.REM_INT_LIT8, Opcode.REM_LONG, Opcode.REM_LONG_2ADDR, Opcode.RSUB_INT, Opcode.RSUB_INT_LIT8, Opcode.SHL_INT, Opcode.SHL_INT_2ADDR, Opcode.SHL_INT_LIT8, Opcode.SHL_LONG, Opcode.SHL_LONG_2ADDR, Opcode.SHR_INT, Opcode.SHR_INT_2ADDR, Opcode.SHR_INT_LIT8, Opcode.SHR_LONG, Opcode.SHR_LONG_2ADDR, Opcode.SUB_DOUBLE, Opcode.SUB_DOUBLE_2ADDR, Opcode.SUB_FLOAT, Opcode.SUB_FLOAT_2ADDR, Opcode.SUB_INT, Opcode.SUB_INT_2ADDR, Opcode.SUB_LONG, Opcode.SUB_LONG_2ADDR, Opcode.USHR_INT, Opcode.USHR_INT_2ADDR, Opcode.USHR_INT_LIT8, Opcode.USHR_LONG, Opcode.USHR_LONG_2ADDR, Opcode.XOR_INT, Opcode.XOR_INT_2ADDR, Opcode.XOR_INT_LIT16, Opcode.XOR_INT_LIT8, Opcode.XOR_LONG, Opcode.XOR_LONG_2ADDR -> BinaryMathOpFactory()
                Opcode.AGET, Opcode.AGET_BOOLEAN, Opcode.AGET_BYTE, Opcode.AGET_CHAR, Opcode.AGET_OBJECT, Opcode.AGET_SHORT, Opcode.AGET_WIDE -> AGetOpFactory()
                Opcode.APUT, Opcode.APUT_BOOLEAN, Opcode.APUT_BYTE, Opcode.APUT_CHAR, Opcode.APUT_OBJECT, Opcode.APUT_SHORT, Opcode.APUT_WIDE -> APutOpFactory()
                Opcode.ARRAY_LENGTH -> ArrayLengthOpFactory()
                Opcode.ARRAY_PAYLOAD -> FillArrayDataPayloadOpFactory()
                Opcode.CHECK_CAST -> CheckCastOpFactory()
                Opcode.CMPG_DOUBLE, Opcode.CMPG_FLOAT, Opcode.CMPL_DOUBLE, Opcode.CMPL_FLOAT, Opcode.CMP_LONG -> CmpOpFactory()
                Opcode.CONST, Opcode.CONST_16, Opcode.CONST_4, Opcode.CONST_CLASS, Opcode.CONST_HIGH16, Opcode.CONST_STRING, Opcode.CONST_STRING_JUMBO, Opcode.CONST_WIDE, Opcode.CONST_WIDE_16, Opcode.CONST_WIDE_32, Opcode.CONST_WIDE_HIGH16 -> ConstOpFactory()
                Opcode.DOUBLE_TO_FLOAT, Opcode.DOUBLE_TO_INT, Opcode.DOUBLE_TO_LONG, Opcode.FLOAT_TO_DOUBLE, Opcode.FLOAT_TO_INT, Opcode.FLOAT_TO_LONG, Opcode.INT_TO_BYTE, Opcode.INT_TO_CHAR, Opcode.INT_TO_DOUBLE, Opcode.INT_TO_FLOAT, Opcode.INT_TO_LONG, Opcode.INT_TO_SHORT, Opcode.LONG_TO_DOUBLE, Opcode.LONG_TO_FLOAT, Opcode.LONG_TO_INT, Opcode.NEG_DOUBLE, Opcode.NEG_FLOAT, Opcode.NEG_INT, Opcode.NEG_LONG, Opcode.NOT_INT, Opcode.NOT_LONG -> UnaryMathOpFactory()
                Opcode.FILLED_NEW_ARRAY, Opcode.FILLED_NEW_ARRAY_RANGE -> FilledNewArrayOpFactory()
                Opcode.FILL_ARRAY_DATA -> FillArrayDataOpFactory()
                Opcode.GOTO, Opcode.GOTO_16, Opcode.GOTO_32 -> GotoOpFactory()
                Opcode.IF_EQ, Opcode.IF_GE, Opcode.IF_GT, Opcode.IF_LE, Opcode.IF_LT, Opcode.IF_NE, Opcode.IF_EQZ, Opcode.IF_GEZ, Opcode.IF_GTZ, Opcode.IF_LEZ, Opcode.IF_LTZ, Opcode.IF_NEZ -> IfOpFactory()
                Opcode.IGET, Opcode.IGET_BOOLEAN, Opcode.IGET_BYTE, Opcode.IGET_CHAR, Opcode.IGET_OBJECT, Opcode.IGET_SHORT, Opcode.IGET_WIDE -> IGetOpFactory()
                Opcode.IPUT, Opcode.IPUT_BOOLEAN, Opcode.IPUT_BYTE, Opcode.IPUT_CHAR, Opcode.IPUT_OBJECT, Opcode.IPUT_SHORT, Opcode.IPUT_WIDE -> IPutOpFactory()
                Opcode.INSTANCE_OF -> InstanceOfOpFactory()
                Opcode.INVOKE_DIRECT, Opcode.INVOKE_INTERFACE, Opcode.INVOKE_STATIC, Opcode.INVOKE_SUPER, Opcode.INVOKE_VIRTUAL, Opcode.INVOKE_DIRECT_RANGE, Opcode.INVOKE_INTERFACE_RANGE, Opcode.INVOKE_STATIC_RANGE, Opcode.INVOKE_SUPER_RANGE, Opcode.INVOKE_VIRTUAL_RANGE -> InvokeOpFactory()
                Opcode.MONITOR_ENTER -> MonitorEnterOpFactory()
                Opcode.MONITOR_EXIT -> MonitorExitOpFactory()
                Opcode.MOVE, Opcode.MOVE_16, Opcode.MOVE_FROM16, Opcode.MOVE_OBJECT, Opcode.MOVE_OBJECT_16, Opcode.MOVE_OBJECT_FROM16, Opcode.MOVE_WIDE, Opcode.MOVE_WIDE_16, Opcode.MOVE_WIDE_FROM16, Opcode.MOVE_EXCEPTION, Opcode.MOVE_RESULT, Opcode.MOVE_RESULT_OBJECT, Opcode.MOVE_RESULT_WIDE -> MoveOpFactory()
                Opcode.NEW_ARRAY -> NewArrayOpFactory()
                Opcode.NEW_INSTANCE -> NewInstanceOpFactory()
                Opcode.NOP -> NopOpFactory()
                Opcode.PACKED_SWITCH, Opcode.SPARSE_SWITCH -> SwitchOpFactory()
                Opcode.PACKED_SWITCH_PAYLOAD, Opcode.SPARSE_SWITCH_PAYLOAD -> SwitchPayloadOpFactory()
                Opcode.RETURN, Opcode.RETURN_WIDE, Opcode.RETURN_OBJECT -> ReturnOpFactory()
                Opcode.RETURN_VOID -> ReturnVoidOpFactory()
                Opcode.SGET, Opcode.SGET_BOOLEAN, Opcode.SGET_BYTE, Opcode.SGET_CHAR, Opcode.SGET_OBJECT, Opcode.SGET_SHORT, Opcode.SGET_WIDE -> SGetOpFactory()
                Opcode.SPUT, Opcode.SPUT_BOOLEAN, Opcode.SPUT_BYTE, Opcode.SPUT_CHAR, Opcode.SPUT_OBJECT, Opcode.SPUT_SHORT, Opcode.SPUT_WIDE -> SPutOpFactory()
                Opcode.THROW -> ThrowOpFactory()
                else -> throw RuntimeException("Unimplemented opcode: $op")
            }
        }
    }
}
